pkg/gadgets/helpers.go

```diff
@@ -187,38 +186,12 @@ func WallTimeFromBootTime(ts uint64) types.Time {
 	return types.Time(time.Unix(0, int64(ts)).Add(timeDiff).UnixNano())
 }
 
-var (
-	bpfKtimeGetBootNsOnce   sync.Once
-	bpfKtimeGetBootNsExists bool
-)
-
-// DetectBpfKtimeGetBootNs returns true if bpf_ktime_get_boot_ns is available
-// in the current kernel. False negatives are possible if BTF is not available.
-func DetectBpfKtimeGetBootNs() bool {
-	bpfKtimeGetBootNsOnce.Do(func() {
-		bpfKtimeGetBootNsExists = false
-
-		btfSpec, err := btf.LoadKernelSpec()
-		if err != nil {
-			return
-		}
-
-		enum := &btf.Enum{}
-		err = btfSpec.TypeByName("bpf_func_id", &enum)
-		if err != nil {
-			return
-		}
-
-		for _, value := range enum.Values {
-			if value.Name == "BPF_FUNC_ktime_get_boot_ns" && value.Value == BpfKtimeGetBootNsFuncID {
-				bpfKtimeGetBootNsExists = true
-
-				return
-			}
-		}
-	})
-
-	return bpfKtimeGetBootNsExists
+// HasBpfKtimeGetBootNs returns true if bpf_ktime_get_boot_ns is available
+func HasBpfKtimeGetBootNs() bool {
+	// We only care about the helper, hence test with ebpf.SocketFilter that exist in all
+	// kernels that support ebpf.
+	err := features.HaveProgramHelper(ebpf.SocketFilter, asm.FnKtimeGetBootNs)
```

Comment:
I wonder about the performance costs of creating a ebpf program just to test the availability of a function, compared to the previous check.

Reply:
I really don't know, previous logic was parsing BTF and looking for a type, that's not a very cheap operation as there are many types there. This approach is loading a program with 3 instructions and not BTF related things at all, it should be pretty fast.  

In any case, this approach is also cached, so the cost is only paid once. Even if this is slower than the previous one, it shouldn't impact the whole performance of Inspektor Gadget at all.

Reply:
We can do that but I don't know if that really makes any difference as the result is already cached by the library.